backend/service/auditsink/slack/slack.go

```diff
@@ -104,3 +121,98 @@ func formatText(username string, event *auditv1.RequestEvent) string {
 
 	return messageText
 }
+
+// FormatCustomText parses out the audit event metdata request for the custom slack message text
+func FormatCustomText(message string, event *auditv1.RequestEvent) (string, error) {
+	tmpl := template.Must(template.New("customText").Funcs(funcMap).Parse(message))
+
+	data, err := getAuditMetadata(event)
+	if err != nil {
+		return "", err
+	}
+
+	var buf bytes.Buffer
+	if err := tmpl.Execute(&buf, data); err != nil {
+		return "", err
+	}
+
+	return buf.String(), nil
+}
+
+// returns the API request/response details contained in a audit event
+func getAuditMetadata(event *auditv1.RequestEvent) (*auditEventMetadata, error) {
+	requestMetadata := event.RequestMetadata.Body
+	// UnmarshalNew returns the message of the specified type based on the TypeURL
+	requestProto, err := requestMetadata.UnmarshalNew()
+	if err != nil {
+		return nil, err
+	}
+
+	responseMetdata := event.ResponseMetadata.Body
+	responseProto, err := responseMetdata.UnmarshalNew()
+	if err != nil {
+		return nil, err
+	}
+
+	return &auditEventMetadata{
+		Request:  requestProto,
+		Response: responseProto,
+	}, nil
+}
+
+func slackList(data interface{}, name string) (string, error) {
```

Comment:
This helper function formats fields that maps and lists & iterates over the values.

I've been stuck/experimenting with how to handle the facet that the map and list values can be any type (usually it's `map<string, string>` or `repeated string` but here's an [example](https://github.com/lyft/clutch/blob/main/api/k8s/v1/k8s.proto#L333) of `map<string, NullableString> where NullableString is a message type with a oneof value).

I currently have something like this https://github.com/lyft/clutch/pull/1093/files#diff-12ec22f81216afa6c8c8d39eeacc01f4ea30bf39b0c3da2525a7f72e7c8beb27R207-R212. The slack output would look like:

the template
```
"*Expected Preconditions*: {{slackList .Request.ExpectedObjectMetaFields `annotations`}}"
```
the slack text
```
`UNKNOWN` performed `UpdatePod` via `clutch.k8s.v1.K8sAPI` using Clutch on resource(s):
- kind-clutch-local/envoy-production/envoy-7784b58d-77m6h (`clutch.k8s.v1.Pod`)
*Expected Preconditions*:
- lyft.net/load-balancing-weight: value:"2"
```

Reply:
I initially was using the reflect package (snippet below) but felt I wasn't taking advantage of the fact that the input is a proto type
```
func slackList(data interface{}) string {
     if data == nil {
             return ""
     }
     switch reflect.TypeOf(data).Kind() {
     case reflect.Slice:
             s := reflect.ValueOf(data)
             for i := 0; i < s.Len(); i++ {
                     text += fmt.Sprintf(sliceFormat, s.Index(i).Interface())
             }
             return text
     case reflect.Map:
             m := reflect.ValueOf(data)
             for _, k := range m.MapKeys() {
                     v := m.MapIndex(k)
                     text += fmt.Sprintf(mapFormat, k.Interface(), v.Interface())
             }
             return text
     }
     return ""
 }
```

Reply:
cc @danielhochman

Reply:
my first inclination is to try to see if we simplify the serialization code by going to JSON first with protojson and using that as the basis for the context. for example, oneofs are flattened back into a top level field by protojson.

i'd need to experiment with it a bit more locally though, maybe you have thoughts on that approach though?